executor/load_data.go

```diff
@@ -608,60 +644,77 @@ func (e *LoadDataInfo) SetMessage() {
 }
 
-func (e *LoadDataInfo) colsToRow(ctx context.Context, cols []field) []types.Datum {
+func (e *LoadDataInfo) colsToRow(ctx context.Context, cols []field, flattenMode bool) (rows [][]types.Datum) {
 	row := make([]types.Datum, 0, len(e.insertColumns))
+handleRemain:
+	for len(cols) > 0 {
+		row = row[:0]
+		var remainIdx int
+		for i := 0; i < len(e.FieldMappings); i++ {
+			if i >= len(cols) {
+				if e.FieldMappings[i].Column == nil {
+					sessionVars := e.Ctx.GetSessionVars()
+					sessionVars.SetUserVar(e.FieldMappings[i].UserVar.Name, "", mysql.DefaultCollationName)
+					continue
+				}
 
-	for i := 0; i < len(e.FieldMappings); i++ {
-		if i >= len(cols) {
+				// If some columns is missing and their type is time and has not null flag, they should be set as current time.
+				if types.IsTypeTime(e.FieldMappings[i].Column.Tp) && mysql.HasNotNullFlag(e.FieldMappings[i].Column.Flag) {
+					row = append(row, types.NewTimeDatum(types.CurrentTime(e.FieldMappings[i].Column.Tp)))
+					continue
+				}
+
+				row = append(row, types.NewDatum(nil))
+				continue
+			}
+
+			remainIdx = i + 1
 			if e.FieldMappings[i].Column == nil {
 				sessionVars := e.Ctx.GetSessionVars()
-				sessionVars.SetUserVar(e.FieldMappings[i].UserVar.Name, "", mysql.DefaultCollationName)
+				sessionVars.SetUserVar(e.FieldMappings[i].UserVar.Name, string(cols[i].str), mysql.DefaultCollationName)
 				continue
 			}
 
-			// If some columns is missing and their type is time and has not null flag, they should be set as current time.
-			if types.IsTypeTime(e.FieldMappings[i].Column.Tp) && mysql.HasNotNullFlag(e.FieldMappings[i].Column.Flag) {
-				row = append(row, types.NewTimeDatum(types.CurrentTime(e.FieldMappings[i].Column.Tp)))
+			// The field with only "\N" in it is handled as NULL in the csv file.
+			// See http://dev.mysql.com/doc/refman/5.7/en/load-data.html
+			if cols[i].maybeNull && string(cols[i].str) == "N" {
+				row = append(row, types.NewDatum(nil))
 				continue
 			}
 
-			row = append(row, types.NewDatum(nil))
-			continue
-		}
-
-		if e.FieldMappings[i].Column == nil {
-			sessionVars := e.Ctx.GetSessionVars()
-			sessionVars.SetUserVar(e.FieldMappings[i].UserVar.Name, string(cols[i].str), mysql.DefaultCollationName)
-			continue
+			row = append(row, types.NewDatum(string(cols[i].str)))
 		}
-
-		// The field with only "\N" in it is handled as NULL in the csv file.
-		// See http://dev.mysql.com/doc/refman/5.7/en/load-data.html
-		if cols[i].maybeNull && string(cols[i].str) == "N" {
-			row = append(row, types.NewDatum(nil))
-			continue
+		if remainIdx >= len(cols) {
+			cols = nil
+		} else {
+			cols = cols[remainIdx:]
+		}
+		for i := 0; i < len(e.ColumnAssignments); i++ {
+			// eval expression of `SET` clause
+			d, err := expression.EvalAstExpr(e.Ctx, e.ColumnAssignments[i].Expr)
+			if err != nil {
+				e.handleWarning(err)
+				rows = append(rows, nil)
+				if !flattenMode {
+					return
+				}
```

Comment:
If it's not `flattenMode` should we return directly here?

Reply:
this keep same with old logic.. https://github.com/pingcap/tidb/pull/22585/files#diff-623a44defe93b702087a0127b2fce250a09a70e1e29840d435c19c865c0e1c5aL652

not flattenMode, should ignore remain columns in current line...

Reply:
A bit confused :frowning:, maybe it needs some refactor in the future.